src/main/include/Constants.h

```diff
@@ -1,154 +1,159 @@
 #pragma once
 
-
-namespace
-GeneralConstants{
-    const double timeStep = 0.02;
-
-    //In meters
-    const double goalHeight = 2.641;
-    const double cameraHeight = 0.4572;
-    const double cameraPitch = 50;
-}
-
-
-namespace
-OIConstants{
-    //Joystick & xbox input ports
-    const int l_joy_port = 0;
-    const int r_joy_port = 1;
-    const int O_joy_port = 2;
+/*
+Coordinate Frames:
+
+   INTAKE
+|O ====== O|
+|          |     +x ^
+|    X     |        |
+|          |  +y    |
+|O        O|  <-----+
+
+
+Angles are measured CCW from +x to +y, per right hand standard.
+X marks the orign of the robot frame. The robot coordinate frame moves with the
+robot. The world frame is equal to the robot frame at t=0, and is fixed.
+*/
+
+#define PI M_PI
+
+namespace GeneralConstants {
+const double timeStep = 0.02;
+
+// In meters
+const double goalHeight = 2.641;
+const double cameraHeight = 0.4572;
+const double cameraPitch = 50;
+}  // namespace GeneralConstants
+
+namespace OIConstants {
+// Joystick & xbox input ports
+const int l_joy_port = 0;
+const int r_joy_port = 1;
+const int O_joy_port = 2;
+}  // namespace OIConstants
+
+namespace DriveConstants {
+// Dimensions of the Robot - SwerveDrive.h
+const double Width = 29;
+const double Length = 29;
+
+// Absolute Encoder offsets
+const double FLOFF = -108.5;
+const double FROFF = 157.0;
+const double BLOFF = 7.91;
+const double BROFF = -18.98;
+
+// Ports for Back Right Swerve Module - SwerveDrive.h
+const int BRanglePort = 18;
+const int BRspeedPort = 17;
+const int BRencoder = 8;
+
+// Ports for Back Left Swerve Module - SwerveDrive.h
+const int BLanglePort = 16;
+const int BLspeedPort = 15;
+const int BLencoder = 42;
+
+// Ports for Front Right Swerve Module - SwerveDrive.h
+const int FRanglePort = 14;
+const int FRspeedPort = 13;
+const int FRencoder = 62;
+
+// Ports for Front Left Swerve Module - SwerveDrive.h
+const int FLanglePort = 12;
+const int FLspeedPort = 11;
+const int FLencoder = 10;
+
+// Wheeldrive
+constexpr double MAX_VOLTS = 1.0;
+const double speedGearRatio = 1 / 6.12;
+const double angleGearRatio = 1 / 12.8;
+const double P = 0.005;
+const double I = 0.001;
+const double D = 0.0002;
+const double Pinit = 0.008;
+const double Iinit = 0.001;
+const double Dinit = 0.0001;
+
+// Swerve Controller
+const double fwdstrP = 0.58;
+const double fwdstrI = 0.0;
+const double fwdstrD = 0.0;
+const double rotP = 0.0011;
+const double rotI = 0.00006;
+const double rotD = 0.0;
+
+const double EMPIRIC_WHEEL_METERS_PER_TICK = 1.0;  // TODO: TUNE

```

Comment:
You need to tune this: setup the robot in teleop with all four swerve wheel encoder values on smartdashboard.
Drive in a straight line with no rotation. Measure the distance the bot traveled in the real world. Take the difference between the before and after encoder values on smart dashboard. Average all four differences, and then divide the number of meters traveled by that average.